lib/common/redis.c

```diff
@@ -38,21 +59,60 @@ static void on_redis_connect(const redisAsyncContext *redis, int status)
         conn->state = H2O_REDIS_CONNECTION_STATE_CLOSED;
         conn->_redis = NULL;
         if (conn->on_close != NULL) {
-            conn->on_close(redis->errstr);
+            conn->on_close(redis->c.errstr);
         }
     }
+
+    if (h2o_timeout_is_linked(&conn->_connect_timeout_entry)) {
+        h2o_timeout_unlink(&conn->_connect_timeout_entry);
+        h2o_timeout_dispose(conn->loop, &conn->_connect_timeout);
+    }
 }
 
-static void on_redis_disconnect(const redisAsyncContext *redis, int status)
+static void close_connection(h2o_redis_conn_t *conn)
 {
-    h2o_redis_conn_t *conn = (h2o_redis_conn_t *)redis->data;
     conn->state = H2O_REDIS_CONNECTION_STATE_CLOSED;
-    conn->_redis = NULL;
+
     if (conn->on_close != NULL) {
-        conn->on_close(redis->errstr);
+        conn->on_close(conn->_redis->c.errstr);
+    }
+    conn->_redis = NULL;
+
+    if (h2o_timeout_is_linked(&conn->_connect_timeout_entry)) {
+        h2o_timeout_unlink(&conn->_connect_timeout_entry);
+    }
+    if (h2o_linklist_is_linked(&conn->_connect_timeout._link)) {
+        h2o_timeout_dispose(conn->loop, &conn->_connect_timeout);
     }
 }
 
+static void on_disconnect(const redisAsyncContext *redis, int status)
+{
+    h2o_redis_conn_t *conn = (h2o_redis_conn_t *)redis->data;
+    close_connection(conn);
+}
+
+static void on_connect_timeout_deferred(h2o_timeout_entry_t *entry)
+{
+    h2o_redis_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_redis_conn_t, _defer_timeout_entry, entry);
+
+    assert((conn->_redis->c.flags & REDIS_CONNECTED) == 0);
+    disconnect(conn, 1);
+
+    /* disconnect function (redisAsyncFree) doesn't call disconnect callback when not the connection is not established */
+    close_connection(conn);
+}
+
+static void on_connect_timeout(h2o_timeout_entry_t *entry)
+{
+    h2o_redis_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_redis_conn_t, _connect_timeout_entry, entry);
+    assert(conn->state != H2O_REDIS_CONNECTION_STATE_CLOSED);
+    h2o_timeout_unlink(entry);
+
+    conn->_did_connect_timeout = 1;
+    invoke_deferred(conn, &conn->_defer_timeout_entry, on_connect_timeout_deferred);
```

Comment:
I would apppreciate if you could rather not make deferred calls within timeout callbacks. I believe that there is no reason to do so. Doing more things at once is generally better for performance (better locality) and debugging (gives us better trace).

Reply:
> I believe that there is no reason to do so.

At that time there're some reasons to do workaround with hiredis async API, but at  https://github.com/h2o/h2o/pull/1152/commits/3ead0dac3b178453c9b592aa3b86fffb3f42a16b I refactored to eliminate such tricks